aiida/backends/sqlalchemy/migrations/versions/d102e96ea943_inline_calc_attributes.py

```diff
@@ -0,0 +1,130 @@
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Migrate attributes from old InlineCalculation that have been transformed into CalcFunctions:
+
+* OLD `sealed` and `function_name` remain the same
+* OLD `namespace` becomes `function_namespace`
+* OLD `first_line_source_code` becomes `function_starting_line_number`
+* OLD `source_file` is moved to repository and dropped from DB
+* OLD `source_code` is used to calculate `function_number_of_lines` and dropped from DB
+* NEW `exit_status` is initialized at 0
+* NEW `process_state` is initialized at "finished"
+* NEW `process_label` is set as `Legacy InlineCalculation`
+
+Note 1: old InlineCalculation can be identified by the old keywords, and after migration they will be
+identified by having `function_number_of_lines` but no `version`.
+
+Note 2: `exit_status` and `process_state` are initialized to 0 and "finished" respectively because the
+CalcFunctions affected are old InlineCalculation, which were only stored if they succeeded. You can check
+the relevant code here (line 60 has the execution call, it is stored after that):
+
+https://github.com/aiidateam/aiida-core/blob/v0.12.5/aiida/orm/implementation/django/calculation/inline.py#L60
+
+Revision ID: d102e96ea943
+Revises: 0edcdd5a30f0
+Create Date: 2021-02-09 16:25:16.106176
+
+"""
+# pylint: disable=invalid-name,no-member
+# yapf: disable
+from alembic import op
+from sqlalchemy.sql import text
+from aiida.backends.general.migrations import utils
+
+# revision identifiers, used by Alembic.
+revision = 'd102e96ea943'
+down_revision = '0edcdd5a30f0'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """Creates the new attributes"""
+    conn = op.get_bind()
+
+    # Creates 'function_namespace' from 'namespace'
+    # Creates 'function_starting_line_number' from 'first_line_source_code'
+    # Creates 'exit_status', 'process_state' and 'process_label'
+    conn.execute(text(
+        """
+        UPDATE db_dbnode
+        SET attributes = attributes || '{
+            "exit_status": 0,
+            "process_state": "finished",
+            "process_label": "Legacy InlineCalculation"
+            }'::jsonb ||
+            jsonb_build_object(
+                'function_starting_line_number', attributes->'first_line_source_code',
+                'function_namespace', attributes->'namespace'
+                )
+        WHERE
+            attributes ? 'namespace' AND
+            attributes ? 'first_line_source_code' AND
+            node_type = 'process.calculation.calcfunction.CalcFunctionNode.';
+        """))
+
+
+    # Sets new 'function_number_of_lines' with the number of lines in 'source_code'
```

Comment:
If I understand correctly, here you are determining the number of lines in the source code by counting the number of new line characters. And to count the new line characters directly in SQL, you are replacing them with empty strings and then compute the character length difference of the new string with the original, which should give the number of lines times the number of characters in the newline character.

It would be good to add this as an explicit comment. Also why you decided to do it this way and not in Python. I could see the point if this would be the only thing you would have to do in Python, but straight after this you are collecting all relevant nodes and writing the source file to the repo. Might as well use the simple Python logic (as in the Django migration) to compute the number of lines. Is a lot less complicated and error prone. Don't think the efficiency would suffer too much.

I am a bit surprised also by the fact that you have to count escaping backslashes in the count. The actual data stored in the database doesn't contain the escape characters does it? So why do you need to count them in the SQL function? Your approach also relies on the fact that you _first_ replace all DOS like endings, then Mac and then Unix. This would also merit a comment because it is not immediately obvious. This is another complexity that would speak in favor of the simply Python approach to be honest

Reply:
> If I understand correctly, here you are determining the number of lines in the source code by counting the number of new line characters. And to count the new line characters directly in SQL, you are replacing them with empty strings and then compute the character length difference of the new string with the original, which should give the number of lines times the number of characters in the newline character.
> 
> It would be good to add this as an explicit comment.
> (...)

But I kind of did, no? See the [two lines below](https://github.com/aiidateam/aiida-core/pull/4750/files/40edeca2e06939f293b5cff2d383e06e6ad73fac#diff-77204f1bd3e424f62e7f81cefb12a090cedaa0188689052fe43957d016144885R75-R76). I mean maybe you mean you want me to explain a bit more?

> Also why you decided to do it this way and not in Python. I could see the point if this would be the only thing you would have to do in Python, but straight after this you are collecting all relevant nodes and writing the source file to the repo. Might as well use the simple Python logic (as in the Django migration) to compute the number of lines. Is a lot less complicated and error prone. Don't think the efficiency would suffer too much.

I had a [first version](https://github.com/aiidateam/aiida-core/pull/4750/commits/ad6efdb1e67a3ad9d9200337a806d05a9d8a4839#diff-77204f1bd3e424f62e7f81cefb12a090cedaa0188689052fe43957d016144885R106-R121) where I did it this way: first I queried for all the `source_code` entries, then I processed those in python and updated the database. The problem with this is that since each update has a specific and different number of lines, I need to do a separate update query for each one of them (thus, the single query to update turns into one query to get the data and n_nodes update queries). Maybe you had a better way in mind that I didn't think of?

Note that for the following section I am doing something similar, yes, but I get all the info for the first query and in the loop I'm just storing the file, there are not queries being made (since I don't need to modify anything that is specific of each node, but just remove the `source_file` attribute once the file is stored, I can just delete all together with the other old attributes in the last query).

I'm honestly not sure what is the performance impact of this: maybe the bottleneck is in the `put_object_from_string` method and then adding a query in each cycle of the same loop does not make a difference. I was under the impression that this is not the case and that each query can take a significant ammount of time, depending on the total size of the DB, but I honestly have made no tests for this. I could try some benchmarking, but if we already have an idea that a single query in a 10MM node DB is comparable to writing a 1k lines file, then I would prefer to accept that and roll with it :sweat_smile:.

> I am a bit surprised also by the fact that you have to count escaping backslashes in the count. The actual data stored in the database doesn't contain the escape characters does it? So why do you need to count them in the SQL function?

Yeah, I didn't understand the why of this either, but I did do some tests to see what was the text stored (because I was getting the wrong count) and indeed the double backlashes were there.

Reply:
> But I kind of did, no? See the two lines below. I mean maybe you mean you want me to explain a bit more?

Sure, this covers the specifics of the solution but it still leaves the reader to cobble together why all of this is needed in the first place. Simply stating something like "Here we need to compute the number of lines in the `source_code` attribute. To be able to do this directly in SQL, we replace all newline characters with empty strings and then compare the new string length with the original to determine the number of characters removed, which corresponds to the number of lines."

Also, I realize now, this solution does not account for a missing terminating newline character. This would give you a count that is one short. Is it guaranteed that the `source_code` always has a terminating newline?

> I had a first version where I did it this way: first I queried for all the source_code entries, then I processed those in python and updated the database. The problem with this is that since each update has a specific and different number of lines, I need to do a separate update query for each one of them 

You do the same in Django though don't you? You get all the nodes and then in a loop update each node, calling `.save()` on it. The same should be possible in SqlAlchemy. Probably it will update it in memory and then at the end perform single commit to the database.

Reply:
> Sure, this covers the specifics of the solution but it still leaves the reader to cobble together why all of this is needed in the first place. Simply stating something like "Here we need to compute the number of lines in the `source_code` attribute. To be able to do this directly in SQL, we replace all newline characters with empty strings and then compare the new string length with the original to determine the number of characters removed, which corresponds to the number of lines."

Ok, gotcha, I will add something along those lines.

> Also, I realize now, this solution does not account for a missing terminating newline character. This would give you a count that is one short. Is it guaranteed that the `source_code` always has a terminating newline?

No, actually I'm assuming it does not have a terminating newline, that is why I always add 1 when storing it:

```SQL
            SET attributes = attributes || jsonb_build_object(
                'function_number_of_lines',
                ( CHAR_LENGTH((attributes -> 'source_code')::TEXT)
                - CHAR_LENGTH( REPLACE((attributes -> 'source_code')::TEXT, :symbol, '' )))
                / :count + 1 -- <<== SEE HERE
                )
```

If it does have a terminating newline then I assume they extra separation line was meant to be considered and add it as well. But I think this starts getting into a terrain that is hard to "homogeneize after the fact". If you really want it, I could add a prior update query that looks for all `source_code` that end in a newline and just remove that, and then apply this, but I don't think I would personally account for that...

> You do the same in Django though don't you? You get all the nodes and then in a loop update each node, calling `.save()` on it. The same should be possible in SqlAlchemy. Probably it will update it in memory and then at the end perform single commit to the database.

This is true, and as you say I did this *with the hope* that the Django ORM is intelligent and knows how to optimize this when building the query since I'm not explicitly saying how to do that. However, I have less confidence that the `conn.execute` method will do this when I'm giving specific SQL queries (I would actually find it a bit concerning if I was explicitly giving queries in SQL and it was deciding to modify them).

As a matter of fact I forgot to do it, but I was actually going to ask if it wouldn't be better to do this part in SQL directly also for django so that we are sure that there is no query bombardment for the ORM not optimizing this request.

Reply:
I am surprised that there is not terminating newline. Maybe the old inline calc code performed a trim on the source code, trimming all leading an trailing whitespace. Would be good if you could check (I think you already looked up the old source to verify that it only got stored on success) and if so then you comment that why it is ok to assume. Then I think it would be fine to keep this SQL solution. I am not 100% sure myself how to do the ORM way through alembic. It is not actually built for data migrations, so maybe it isn't possible. For this item would still like @giovannipizzi to also have a look. Know he is busy, but I don't feel comfortable signing off on migrations that I am not fully sure of.

Reply:
> I am surprised that there is not terminating newline. 

Ok, I think I may have been unclear on my previous statement, let me try to clarify. My assumption is not on the existence on the ending newline, but on the consistency of how this used to be stored and how we want the soon to be counting to work. Let me explain:

I don't know how AiiDA stored the `source_code` in the attributes, but if we use to store the ending newline I assume we would also want that consider in the new count feature (since it is mostly used internally to return the same content but via a method that takes it from the file instead of keeping it in the DB).

I can try looking in the some of the old DB of the mat cloud for how this was stored, if they had the ending newline or not, and report back (althogh would that be enough to assume it was always the same?). However, independent of confirming its existence, it doesn't really tell me wether I should count or not: I don't know how @unkcpz or you were conceiving this count to work. Again, my assumption was that the new method would return the exact same thing that we use to store in the DB.

Reply:
> but if we use to store the ending newline I assume we would also want that consider in the new count feature (since it is mostly used internally to return the same content but via a method that takes it from the file instead of keeping it in the DB).

The point is though that in the new version we no longer store the source code lines of the function, but just the number of lines in the source code. Together with the starting line, returning the exact lines of the function is then well-defined. The only question you might have there is whether the string that is returned has the terminating newline or not, but that can be controlled on the fly.

Here, the question I am asking, is whether the calculation you are making to determine the number of lines in the functions source code based on the string in the database is correct. You can assume that it was at least always created in the same way and so if the current method works, it would always work, but that is clearly an assumption and can actually fail. If originally the following was stored:
```
source_code = 'def some_function(arg):\n    return\n
```
but then that was changed to trim the string before storing it such that 
```
source_code = 'def some_function(arg):\n    return
```
was stored instead, your code would fail for certain nodes. For the latter it would correctly return 2 lines, but for the former it would return 3. Now luckily, in this case, the new code would return all the source code _plus an additional line_ from the source file, which is likely to be an empty line, so it won't be a problem. But still, this is the problem I was describing and why I was arguing to instead do this in Python where accounting for both cases is trivial.

Reply:
I couldn't follow the whole discussion on how to exactly count the number of lines, I have to admit. I trust you did the right counting :-D

I however think that, indeed, it's better to try to keep the same SQL code also in Django, maybe, so we are sure we're doing the very same thing in both backends? Especially in this case, where the logic is *so* different. 
Also in the future, if we have similar issues, we'll need to do it anyway also for SQLA, so I don't think it's a big advantage to keep the Django syntax in the codebase.
On the efficiency, as I mention earlier: every save will trigger a commit to the DB (still within the transaction, if you are in one). So many commits are slower. I cannot say if the time for a commit and the time to save a file are comparable or one is much larger than the other, so I cannot say if moving all in a loop helps or not.
However, the weird substitution is quite complex to understand I have to admit... I cannot say by looking at it if it's correct - I trust that we have enough tests that verify it...